### PR TITLE
Add support for RSA SHA-2 public key algorithms

### LINKED ISSUE #825: Support for rsa-sha2-256 and rsa-sha2-512
Do you have plan to support for rsa-sha2-256 and rsa-sha2-512?

### LINKED ISSUE #956: Permission denied (publickey) with SSH.NET - Can connect via FileZilla or Powershell.
I'm trying to extend our SFTP capability to allow connection via key-authentication.

The code examples are very straightforward (thanks for that) and I'm pretty sure I've done it right:
`connectionInfo = New ConnectionInfo(e.Host, e.Port, e.Username, GetAuthenticationMethods(e))`
with GetAuthenticationMethods(e) giving me an array of AuthenticationMethods, depending on the provided information:

```
[...]
authMethod.Add(New PasswordAuthenticationMethod(e.Username, e.Password))
[...]
If e.FilePassphrase = "" Then
  authMethod.Add(New PrivateKeyAuthenticationMethod(e.Username, New PrivateKeyFile(keyFile)))
ElseIf e.FilePassphrase <> "" Then
  authMethod.Add(New PrivateKeyAuthenticationMethod(e.Username, New PrivateKeyFile(keyFile, e.FilePassphrase)))
End If
[...]
m_Client = New SftpClient(connectionInfo)
m_Client.Connect()
```

I have a Server set up and generated keys. Originally, I had a ssh-rsa key, giving me a `openssh key type: ssh-rsa is not supported` Error while trying to instantiate the PrivateKeyAuthenticationMethod. (afaict this should be supported but is not yet in the NuGet package, is there any ETA on when this will be available?)

I converted this key to PEM, which gets me one step further but now I get a  `Permission denied (publickey).` Error. 

I can connect to the server without issues using either the PEM or the ssh-rsa key over FileZilla or ssh on console.

### LINKED ISSUE #1105: Key exchange negotiation failed
Renci.SshNet.Common.SshConnectionException: Key exchange negotiation failed.
   at Renci.SshNet.Session.WaitOnHandle(WaitHandle waitHandle, TimeSpan timeout) in \SSH.NET-rsa-sha2-256\src\Renci.SshNet\Session.cs:line 977
   at Renci.SshNet.Session.WaitOnHandle(WaitHandle waitHandle) in \SSH.NET-rsa-sha2-256\src\Renci.SshNet\Session.cs:line 874
   at Renci.SshNet.Session.ConnectAsync(CancellationToken cancellationToken) in \SSH.NET-rsa-sha2-256\src\Renci.SshNet\Session.cs:line 739
   at Renci.SshNet.BaseClient.CreateAndConnectSessionAsync(CancellationToken cancellationToken) in \SSH.NET-rsa-sha2-256\src\Renci.SshNet\BaseClient.cs:line 545
   at Renci.SshNet.BaseClient.ConnectAsync(CancellationToken cancellationToken) in \SSH.NET-rsa-sha2-256\src\Renci.SshNet\BaseClient.cs:line 284
   at

### LINKED ISSUE #1116: Error 'Permission denied (publickey).' while connecting to AWS SSH
Renci.SshNet.Common.SshAuthenticationException: 'Permission denied (publickey).' occurs while connecting to Amazon Linux AWS instance. 
SSH and Putty connect to this server without any problem.
Client OS - Windows 10 x64, .net6, ssh.net 2022.0.2
ssh output:
```
ssh.exe -vvv -i "key.pem" ec2-user@ec2-xxx.amazonaws.com

OpenSSH_9.0p1, OpenSSL 1.1.1p  21 Jun 2022
debug1: Reading configuration data /etc/ssh/ssh_config
debug3: expanded UserKnownHostsFile '~/.ssh/known_hosts' -> '/c/Users/xxx/.ssh/known_hosts'
debug3: expanded UserKnownHostsFile '~/.ssh/known_hosts2' -> '/c/Users/xxx/.ssh/known_hosts2'
debug2: resolving "ec2-xxx.amazonaws.com" port 22
debug3: resolve_host: lookup ec2-xxx.amazonaws.com:22
debug3: ssh_connect_direct: entering
debug1: Connecting to ec2-xxx.amazonaws.com [54.xxx.240.111] port 22.
debug3: set_sock_tos: set socket 4 IP_TOS 0x48
debug1: Connection established.
debug1: identity file E:\\Amazon\\key.pem type -1
debug1: identity file E:\\Amazon\\key.pem-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_9.0
debug1: Remote protocol version 2.0, remote software version OpenSSH_8.7
debug1: compat_banner: match: OpenSSH_8.7 pat OpenSSH* compat 0x04000000
debug2: fd 4 setting O_NONBLOCK
debug1: Authenticating to ec2-xxx.amazonaws.com:22 as 'ec2-user'
debug3: record_hostkey: found key type ED25519 in file /c/Users/xxx/.ssh/known_hosts:4
debug3: record_hostkey: found key type ECDSA in file /c/Users/xxx/.ssh/known_hosts:5
debug3: load_hostkeys_file: loaded 2 keys from ec2-xxx.amazonaws.com
debug1: load_hostkeys: fopen /c/Users/xxx/.ssh/known_hosts2: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts2: No such file or directory
debug3: order_hostkeyalgs: have matching best-preference key type ssh-ed25519-cert-v01@openssh.com, using HostkeyAlgorithms verbatim
debug3: send packet: type 20
debug1: SSH2_MSG_KEXINIT sent
debug3: receive packet: type 20
debug1: SSH2_MSG_KEXINIT received
debug2: local client KEXINIT proposal
debug2: KEX algorithms: sntrup761x25519-sha512@openssh.com,curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256,ext-info-c
debug2: host key algorithms: ssh-ed25519-cert-v01@openssh.com,ecdsa-sha2-nistp256-cert-v01@openssh.com,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp521-cert-v01@openssh.com,sk-ssh-ed25519-cert-v01@openssh.com,sk-ecdsa-sha2-nistp256-cert-v01@openssh.com,rsa-sha2-512-cert-v01@openssh.com,rsa-sha2-256-cert-v01@openssh.com,ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,sk-ssh-ed25519@openssh.com,sk-ecdsa-sha2-nistp256@openssh.com,rsa-sha2-512,rsa-sha2-256
debug2: ciphers ctos: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
debug2: ciphers stoc: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
debug2: MACs ctos: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
debug2: MACs stoc: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
debug2: compression ctos: none,zlib@openssh.com,zlib
debug2: compression stoc: none,zlib@openssh.com,zlib
debug2: languages ctos:
debug2: languages stoc:
debug2: first_kex_follows 0
debug2: reserved 0
debug2: peer server KEXINIT proposal
debug2: KEX algorithms: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512
debug2: host key algorithms: ecdsa-sha2-nistp256,ssh-ed25519
debug2: ciphers ctos: aes256-gcm@openssh.com,chacha20-poly1305@openssh.com,aes256-ctr,aes128-gcm@openssh.com,aes128-ctr
debug2: ciphers stoc: aes256-gcm@openssh.com,chacha20-poly1305@openssh.com,aes256-ctr,aes128-gcm@openssh.com,aes128-ctr
debug2: MACs ctos: hmac-sha2-256-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha2-256,hmac-sha1,umac-128@openssh.com,hmac-sha2-512
debug2: MACs stoc: hmac-sha2-256-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha2-256,hmac-sha1,umac-128@openssh.com,hmac-sha2-512
debug2: compression ctos: none,zlib@openssh.com
debug2: compression stoc: none,zlib@openssh.com
debug2: languages ctos:
debug2: languages stoc:
debug2: first_kex_follows 0
debug2: reserved 0
debug1: kex: algorithm: curve25519-sha256
debug1: kex: host key algorithm: ssh-ed25519
debug1: kex: server->client cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug1: kex: client->server cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug3: send packet: type 30
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug3: receive packet: type 31
debug1: SSH2_MSG_KEX_ECDH_REPLY received
debug1: Server host key: ssh-ed25519 SHA256:r7Aez12NhVjT4TaPMPLO+QrTWm5Mwi1Eiis3lgjxcE4
debug3: record_hostkey: found key type ED25519 in file /c/Users/xxx/.ssh/known_hosts:4
debug3: record_hostkey: found key type ECDSA in file /c/Users/xxx/.ssh/known_hosts:5
debug3: load_hostkeys_file: loaded 2 keys from ec2-xxx.amazonaws.com
debug1: load_hostkeys: fopen /c/Users/xxx/.ssh/known_hosts2: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts2: No such file or directory
debug1: Host 'ec2-xxx.amazonaws.com' is known and matches the ED25519 host key.
debug1: Found key in /c/Users/xxx/.ssh/known_hosts:4
debug3: send packet: type 21
debug2: ssh_set_newkeys: mode 1
debug1: rekey out after 134217728 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug3: receive packet: type 21
debug1: SSH2_MSG_NEWKEYS received
debug2: ssh_set_newkeys: mode 0
debug1: rekey in after 134217728 blocks
debug1: Will attempt key: key.pem  explicit
debug2: pubkey_prepare: done
debug3: send packet: type 5
debug3: receive packet: type 7
debug1: SSH2_MSG_EXT_INFO received
debug1: kex_input_ext_info: server-sig-algs=<ssh-ed25519,sk-ssh-ed25519@openssh.com,ssh-rsa,rsa-sha2-256,rsa-sha2-512,ssh-dss,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,sk-ecdsa-sha2-nistp256@openssh.com,webauthn-sk-ecdsa-sha2-nistp256@openssh.com>
debug3: receive packet: type 6
debug2: service_accept: ssh-userauth
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug3: send packet: type 50
debug3: receive packet: type 51
debug1: Authentications that can continue: publickey,gssapi-keyex,gssapi-with-mic
debug3: start over, passed a different list publickey,gssapi-keyex,gssapi-with-mic
debug3: preferred publickey,keyboard-interactive,password
debug3: authmethod_lookup publickey
debug3: remaining preferred: keyboard-interactive,password
debug3: authmethod_is_enabled publickey
debug1: Next authentication method: publickey
debug1: Trying private key: key.pem
debug3: sign_and_send_pubkey: using publickey with RSA SHA256:05GwOkwa74XFyzi+37mNV1oq9uagMv+IAj8OJ1eyeVU
debug3: sign_and_send_pubkey: signing using rsa-sha2-512 SHA256:05GwOkwa74XFyzi+37mNV1oq9uagMv+IAj8OJ1eyeVU
debug3: send packet: type 50
debug2: we sent a publickey packet, wait for reply
debug3: receive packet: type 52
Authenticated to ec2-xxx.amazonaws.com ([54.xxx.240.111]:22) using "publickey".
debug1: channel 0: new [client-session]
debug3: ssh_session2_open: channel_new: 0
debug2: channel 0: send open
debug3: send packet: type 90
debug1: Requesting no-more-sessions@openssh.com
debug3: send packet: type 80
debug1: Entering interactive session.
debug1: pledge: filesystem
debug3: receive packet: type 80
debug1: client_input_global_request: rtype hostkeys-00@openssh.com want_reply 0
debug3: client_input_hostkeys: received ECDSA key SHA256:NcWBX8ZbscSZ3LL1vawtZ7Kzu7zpOa2DspdROVdPYYU
debug3: client_input_hostkeys: received ED25519 key SHA256:r7Aez12NhVjT4TaPMPLO+QrTWm5Mwi1Eiis3lgjxcE4
debug1: client_input_hostkeys: searching /c/Users/xxx/.ssh/known_hosts for ec2-xxx.amazonaws.com / (none)
debug3: hostkeys_foreach: reading file "/c/Users/xxx/.ssh/known_hosts"
debug3: hostkeys_find: found ssh-ed25519 key at /c/Users/xxx/.ssh/known_hosts:4
debug3: hostkeys_find: found ecdsa-sha2-nistp256 key at /c/Users/xxx/.ssh/known_hosts:5
debug1: client_input_hostkeys: searching /c/Users/xxx/.ssh/known_hosts2 for ec2-xxx.amazonaws.com / (none)
debug1: client_input_hostkeys: hostkeys file /c/Users/xxx/.ssh/known_hosts2 does not exist
debug3: client_input_hostkeys: 2 server keys: 0 new, 2 retained, 0 incomplete match. 0 to remove
debug1: client_input_hostkeys: no new or deprecated keys from server
debug3: receive packet: type 4
debug1: Remote: /home/ec2-user/.ssh/authorized_keys:1: key options: agent-forwarding port-forwarding pty user-rc x11-forwarding
debug3: receive packet: type 91
debug2: channel_input_open_confirmation: channel 0: callback start
debug2: fd 4 setting TCP_NODELAY
debug3: set_sock_tos: set socket 4 IP_TOS 0x48
debug2: client_session2_setup: id 0
debug2: channel 0: request pty-req confirm 1
debug3: send packet: type 98
debug2: channel 0: request shell confirm 1
debug3: send packet: type 98
debug2: channel_input_open_confirmation: channel 0: callback done
debug2: channel 0: open confirm rwindow 0 rmax 32768
debug3: receive packet: type 99
debug2: channel_input_status_confirm: type 99 id 0
debug2: PTY allocation request accepted on channel 0
debug2: channel 0: rcvd adjust 2097152
debug3: receive packet: type 99
debug2: channel_input_status_confirm: type 99 id 0
debug2: shell request accepted on channel 0
   ,     #_
   ~\_  ####_        Amazon Linux 2023
  ~~  \_#####\
  ~~     \###|
  ~~       \#/ ___   https://aws.amazon.com/linux/amazon-linux-2023
   ~~       V~' '->
    ~~~         /
      ~~._.   _/
         _/ _/
       _/m/'
Last login: Mon May  8 10:51:14 2023 from 185.xxx.155.152
[ec2-user@ip-172-xxx-92-171 ~]$
```

### LINKED ISSUE #1145: 'Permission denied (publickey).' but terminal ssh connection works fine
Hey,

Maybe i am missing the obvious here, but i see no reason why my connection should not work.

The Key is already converted into something SSH.net usable:

```
-----BEGIN RSA PRIVATE KEY-----
Proc-Type: 4,ENCRYPTED
DEK-Info: AES-128-CBC,

...

-----END RSA PRIVATE KEY-----
```
and a corresponding ssh call shows the public key is registered on the server correctly:

```
ssh DOMAIN -l root -i C:\Users\XXXX\.ssh\id_XXXX
-> Works
```

Then why does this code throws an exception on connect?

```csharp
var connectionInfo = new ConnectionInfo(uploadTargetHost,
    sshUser,
    new PrivateKeyAuthenticationMethod(sshUser, new PrivateKeyFile(@"C:\Users\XXXX\.ssh\id_XXXX", keyPassword)));

using var client = new SftpClient(connectionInfo);
EventHandler<HostKeyEventArgs> clientOnHostKeyReceived = (sender, e) =>
{
    e.CanTrust = true;
};
client.HostKeyReceived += clientOnHostKeyReceived;
client.Connect();
```

```
Renci.SshNet.Common.SshAuthenticationException: 'Permission denied (publickey).'
```

### LINKED ISSUE #1174: RSA SHA-2 implementation proposal
## Background
The public key algorithm `ssh-rsa` which uses RSA and the SHA-1 hash algorithm, is considered deficient. OpenSSH version 8.8 (released on 2021-09-26) disables `ssh-rsa` by default. See [RFC 8332](https://datatracker.ietf.org/doc/html/rfc8332) and the [OpenSSH 8.8 release notes](https://www.openssh.com/txt/release-8.8).

This library does not currently support the stronger public key algorithms using RSA with SHA-2 (`rsa-sha2-256` and `rsa-sha2-512`) resulting in `Host key algorithm negotiation failed` or similar. It is probably the number one problem by number of issues and comments.

There is a workaround/solution provided in #825, #971 (and subsequently #1156). The workaround creates new `RsaKey` and `RsaDigitalSignature` classes for SHA-256 (and presumably more for SHA-512). The duplication is made understandably due to the library's coupling between the `Key` and `DigitalSignature` types (see also https://github.com/sshnet/SSH.NET/pull/1156#issuecomment-1672002953). The coupling is also understandable because until now one type of key has only had one type of signature. However, the workaround is IMO not adequate for the library implementation.

This issue is a proposal for the implementation. I have opened PR #1170 to allow specifying the hash algorithm in the `RsaDigitalSignature` constructor. This proposal builds on that.

## Problem 1: specifying the hash algorithm when instantiating a `KeyHostAlgorithm`
The `KeyHostAlgorithm` type performs signing/verification and (de)serialisation of the data to the binary message format. A `KeyHostAlgorithm` takes a `Key`, and the signature is performed via the `Key.Sign` and `Key.VerifySignature` methods, which are non-virtual and simply forward to the `DigitalSignature.Sign` and `DigitalSignature.Verify` methods. Because `KeyHostAlgorithm` just accepts a `Key`, there is no way to specify the signature type.

### Solution 1: add `RsaKey` constructors taking a factory method
We add two new constructors to `RsaKey` taking a `Func<RsaKey, RsaDigitalSignature>` parameter. When the `DigitalSignature` getter is invoked, if the delegate has been supplied then it will be invoked with a parameter of `this`, and the `RsaDigitalSignature` returned from the delegate is returned from the getter. If the delegate has not been supplied, the implementation will keep its current behaviour of instantiating a SHA-1 `RsaDigitalSignature`.

```diff
public class RsaKey : Key, IDisposable
{
    // Existing
    public RsaKey();
    public RsaKey(byte[] data);
    public RsaKey(BigInteger modulus, BigInteger exponent, BigInteger d, BigInteger p, BigInteger q, BigInteger inverseQ)

    // Proposed
+   public RsaKey(Func<RsaKey, RsaDigitalSignature> digitalSignatureFactory);
+   public RsaKey(byte[] data, Func<RsaKey, RsaDigitalSignature> digitalSignatureFactory);
}
```

#### Usage
```c#
KeyHostAlgorithm alg = new("rsa-sha2-256", new RsaKey(key => new RsaDigitalSignature(key, HashAlgorithmName.SHA256)));
```

#### Advantages
* No breaking changes
* The changes are limited to RSA i.e. only where it is needed
#### Disadvantages
* The `RsaKey` type is concerned with the implementation of the signature. (Although it is already)

### Solution 2: supply a `DigitalSignature` instance to `KeyHostAlgorithm`
By adding a settable `DigitalSignature` property or a new constructor parameter to `KeyHostAlgorithm`, we can allow specifying the signature to be used.

#### Usage

```c#
RsaKey key = new();
KeyHostAlgorithm alg = new("rsa-sha2-256", key, new RsaDigitalSignature(key, HashAlgorithmName.SHA256));
```

#### Advantages
* Non-breaking, we can continue to sign via the `Key` if a `DigitalSignature` has not been supplied
#### Disadvantages
* We cannot be certain that the supplied `DigitalSignature` uses the supplied `Key`, which may be a source of bugs and issues.
* Having two ways to perform the signature (via the `Key` and via the `DigitalSignature`) may be a source of confusion.

### Solution 3: decouple `Key` and `DigitalSignature`
We remove the signature functionality from `Key` and obsolete its `DigitalSignature` property. `KeyHostAlgorithm` would then gain a `DigitalSignature` property and constructors taking a `DigitalSignature` instead of a `Key`. `DigitalSignature` would gain a `Key` property. The old constructors would be obsoleted akin to:

```diff
+[Obsolete("Use the KeyHostAlgorithm(string, DigitalSignature) ctor in order to specify the digital signature")]
public KeyHostAlgorithm(string name, Key key)
    : base(name)
{
    Key = key;
+   // Key.DigitalSignature is changed from 'protected abstract' to 'protected internal virtual'
+   DigitalSignature = key.DigitalSignature
+       ?? throw new InvalidOperationException($"No {nameof(DigitalSignature)} found. This constructor is obselete.");
}
```

#### Usage

```c#
RsaKey key = new();
KeyHostAlgorithm alg = new("rsa-sha2-256", new RsaDigitalSignature(key, HashAlgorithmName.SHA256));
```

#### Advantages
* The two-way, chicken-and-egg dependency between `Key` and `DigitalSignature` no longer exists
#### Disadvantages
* *Big* breaking change
* The changes would apply unnecessarily to all key algorithms, not just RSA

## Problem 2: specifying multiple host algorithms for one private key
An RSA private key should be able to be used for multiple host key algorithms (`ssh-rsa`, `rsa-sha2-256` and `rsa-sha2-512`). Currently, `PrivateKeyFile` implements the `IPrivateKeySource` interface:

```c#
public interface IPrivateKeySource
{
    HostAlgorithm HostKey { get; }
}
```

which is what's supplied and enumerated during authentication:

https://github.com/sshnet/SSH.NET/blob/9f1699b5602807e50b8ea6b4300a12ff9d00f93f/src/Renci.SshNet/PrivateKeyAuthenticationMethod.cs#L41

### Solution 1: introduce a new interface
We add a new interface which can provide multiple host algorithms:

```diff
// See comments on naming below
+public interface IExtendedPrivateKeySource : IPrivateKeySource
+{
+   IReadOnlyList<HostAlgorithm> HostKeys { get; }
+}

-public class PrivateKeyFile : IPrivateKeySource, IDisposable
+public class PrivateKeyFile : IExtendedPrivateKeySource, IDisposable
{ }
```

The enumeration of `IPrivateKeySource`s during authentication would make a check for whether the instance is an `IExtendedPrivateKeySource` and then enumerate its `HostKeys` too. The `HostKeys` property is an `IReadOnlyList` to allow counting and indexing such that the `HostAlgorithm`s can be supplied in order of preference. `IExtendedPrivateKeySource` inherits `IPrivateKeySource` because `PrivateKeyAuthenticationMethod` would be the only place it is used in the library, meaning every implementation of `IExtendedPrivateKeySource` would have to implement `IPrivateKeySource` anyway. `IPrivateKeySource.HostKey` would generally defer to `IExtendedPrivateKeySource .HostKeys[0]` representing the most preferred algorithm. 

On naming, the `IPrivateKeySource` is IMO named slightly awkwardly: functionally, it is an abstraction of something which provides a `HostAlgorithm`, with no explicit relevance to a private key. And if we wanted such an abstraction for a different use, we would either be using this interface having an unrelated name, or an interface with the same definition but a different name. Both of these would be strange.

The interface was introduced in bc99ada7 (March 22) meaning it does not exist in the latest released package, but there were a number of requests for it so it is safest to assume that it is in use in local builds.

While I would strongly prefer not to make any breaking changes in this feature, I would be comfortable with an exception to rename `IPrivateKeySource` to something more relevant, like `IHostAlgorithmProvider` with an obsoletion message at error severity, before the next release. The new interface proposed above would be called `IHostAlgorithmsProvider`. This solution 1 includes the rename.

#### Advantages
* Suitable, future-proof names for interfaces
#### Disadvantages
* Breaks people implementing `IPrivateKeySource` in local builds
* Inheriting `IPrivateKeySource` in the new interface may be awkward and unnecessary (I do not feel strongly about it)

### Solution 1b: the same as solution 1 but without renaming `IPrivateKeySource`
We should then decide whether to keep the new interface in line with the naming e.g. `IExtendedPrivateKeySource` or go the more generic route of `IHostAlgorithmsProvider`.

#### Advantages
* Breaks nobody
#### Disadvantages
* We are left with strangely named interface(s) which do not match their function

### Solution 2: add `HostKeys` property to `PrivateKeyFile` only
And then during authentication, check if the `IPrivateKeySource` is a concrete `PrivateKeyFile` and enumerate its `HostKeys`

#### Disadvantages
* Defeats the purpose of the interface, which was well-requested.

### Solution 3: change the `IPrivateKeySource` interface
Change the `HostKey` property to a collection, or add the collection property directly to the interface.

#### Advantages
* Avoids the need for another interface
#### Disadvantages
* Breaks people implementing `IPrivateKeySource` in local builds (even more than a rename)

---

Thanks for reading down to here. My personal preference is for solution 1 to problem 1, and solution 1a to problem 2. I would be willing to implement that relatively quickly.

### PR DESCRIPTION
**Pull request by Rob-Hague**
See #1174 and [RFC 8332](https://datatracker.ietf.org/doc/html/rfc8332) for details.

closes #1174 
closes #1145 
closes #1116
closes #1105
closes #956
closes #825

It may also close the following:
#1141
#1025
#1003
#858
#845

### CODE REVIEW

**Comment by WojciechNagorski** [comment]
@Rob-Hague I've pushed some changes to this PR. Can you review this?

File: src/Renci.SshNet/PrivateKeyAuthenticationMethod.cs [hunk 1]
```diff
@@ -111,7 +111,7 @@ public override AuthenticationResult Authenticate(Session session)
 
                     session.WaitOnHandle(_authenticationCompleted);
 
-                    if (_authenticationResult == AuthenticationResult.Success)
+                    if (_authenticationResult is AuthenticationResult.Success or AuthenticationResult.PartialSuccess)
```
Comment (Rob-Hague) [inline src/Renci.SshNet/PrivateKeyAuthenticationMethod.cs:114]: I *think* this change makes sense. Was there something not working which motivated the change?

**Comment by Rob-Hague** [comment]
Thanks! I should have paid more attention to `PrivateKeyFile`. I will add test coverage for that. And I will review the other changes more closely and let you know.

File: src/Renci.SshNet/PrivateKeyAuthenticationMethod.cs (hunk 1 above)
Comment (WojciechNagorski) [reply src/Renci.SshNet/PrivateKeyAuthenticationMethod.cs:114]: You can remove this line locally which will cause one test to fail. This happened when I removed PubkeyAcceptedAlgorithms from the config. I think this change is okay, but I'd love it if you checked out this line yourself

**Comment by WojciechNagorski** [comment]
Thanks to removing PubkeyAcceptedAlgorithms from the config file we know that SSH.NET supports the latest version of OpenSSH without any hack (configuration).

File: src/Renci.SshNet/IPrivateKeySource.cs [hunk 2]
```diff
@@ -1,15 +1,22 @@
...
 {
     /// <summary>
     /// Represents private key source interface.
     /// </summary>
-    public interface IPrivateKeySource
+    /// <remarks>
+    /// This interface has been replaced by <see cref="IHostAlgorithmsProvider"/>
+    /// and is currently not used in the library.
+    /// </remarks>
+    [EditorBrowsable(EditorBrowsableState.Never)]
```
Comment (WojciechNagorski) [inline src/Renci.SshNet/IPrivateKeySource.cs:18]: Maybe we should add [Obsolete("comment")]?

File: src/Renci.SshNet/Security/Cryptography/RsaDigitalSignature.cs [hunk 3]
```diff
@@ -14,13 +13,23 @@ public class RsaDigitalSignature : CipherDigitalSignature, IDisposable
         private HashAlgorithm _hash;
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="RsaDigitalSignature"/> class.
+        /// Initializes a new instance of the <see cref="RsaDigitalSignature"/> class with the SHA-1 hash algorithm.
         /// </summary>
         /// <param name="rsaKey">The RSA key.</param>
         public RsaDigitalSignature(RsaKey rsaKey)
```
Comment (WojciechNagorski) [inline src/Renci.SshNet/Security/Cryptography/RsaDigitalSignature.cs:19]: how about adding `[Obsolete("Comment")]` and using always new constructor?

File: src/Renci.SshNet/PrivateKeyAuthenticationMethod.cs (hunk 1 above)
Comment (Rob-Hague) [reply src/Renci.SshNet/PrivateKeyAuthenticationMethod.cs:114]: I am not super familiar with multi-factor auth in SSH, but the reason for the failure and the fix both make sense to me. Thanks

File: src/Renci.SshNet/IPrivateKeySource.cs (hunk 2 above)
Comment (Rob-Hague) [reply src/Renci.SshNet/IPrivateKeySource.cs:18]: Done. I initially could not decide but agree it is the better thing to do

File: src/Renci.SshNet/Security/Cryptography/RsaDigitalSignature.cs (hunk 3 above)
Comment (Rob-Hague) [reply src/Renci.SshNet/Security/Cryptography/RsaDigitalSignature.cs:19]: I do not think it is necessary here

(but I don't mind)

**Comment by WojciechNagorski** [comment]
It looks that everything is ready. I have to look over `IHostAlgorithmsProvider` `IPrivateKeySource`

**Review by WojciechNagorski — APPROVED** [review]
Great work! Without you, we definitely wouldn't have rsa-sha-256 and rsa-sha-512 in this release. I'm merging this PR as is, but I will change the approach for `IHostAlgorithmsProvide` in the next PR. I've done many tests and experiments and it turned out that this was not the best approach.

**Comment by Rob-Hague** [comment]
Awesome. Your approach to `IPrivateKeySource` is better than mine, I was just overly cautious about breaking changes. Thanks

**Comment by et1975** [comment]
When are you thinking to release?

**Comment by WojciechNagorski** [comment]
This week or next.

**Comment by dominikamann** [comment]
@WojciechNagorski I am very grateful about this release :) It will help us a lot. Thank you!

**Comment by dominikamann** [comment]
I tested it with the current development branch. But it is still not working for me. Over console window with ssh root@xyz it works fine.

grafik

Exception in ClientAuthentication.cs:
grafik

Do you have any Idea? 

Thank you all!

**Comment by Rob-Hague** [comment]
Can you step through `PrivateKeyAuthenticationMethod.Authenticate` to see what algorithms are being used?

What are the algorithms returned in `KeyExchangeInitMessage.Process`?

edit: your key path also says ed25519. Are you sure this is related to rsa?

**Comment by dominikamann** [comment]
Thank you ... sure

grafik

grafik

grafik

**Comment by Rob-Hague** [comment]
Your key is an ed25519 so I would suggest that it is not related to this PR.

Would you mind opening another issue and pasting the result of ssh console command with "-v" (verbose) argument?

**Comment by YohanSciubukgian** [comment]
> This week or next.

Do you have any update to share on the delivery timeline please ?

**Comment by WojciechNagorski** [comment]
@YohanSciubukgian 
I hope so that it will work. I'm waiting for one thing from @drieseng.

If it doesn't work out this week, I'll do it at the beginning of next week, regardless of the thing I'm waiting for.

**Pull request merged**